common/http/platform/beast/http.cpp

```diff
@@ -0,0 +1,366 @@
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <common/common.hpp>
+#include <common/http.hpp>
+
+namespace mender {
+namespace http {
+
+namespace common = mender::common;
+
+// At the time of writing, Beast only supports HTTP/1.1, and is unlikely to support HTTP/2 according
+// to this discussion: https://github.com/boostorg/beast/issues/1302.
+const unsigned int BeastHttpVersion = 11;
+
+namespace asio = boost::asio;
+namespace http = boost::beast::http;
+
+static http::verb VerbToBeastVerb(Verb verb) {
+	switch (verb) {
+	case GET:
+		return http::verb::get;
+	case POST:
+		return http::verb::post;
+	case PUT:
+		return http::verb::put;
+	case PATCH:
+		return http::verb::patch;
+	case CONNECT:
+		return http::verb::connect;
+	}
+	// Don't use "default" case. This should generate a warning if we ever add any verbs. But
+	// still assert here for safety.
+	assert(false);
+	return http::verb::get;
+}
+
+static Verb BeastVerbToVerb(http::verb verb) {
+	switch (verb) {
+	case http::verb::get:
+		return GET;
+	case http::verb::post:
+		return POST;
+	case http::verb::put:
+		return PUT;
+	case http::verb::patch:
+		return PATCH;
+	case http::verb::connect:
+		return CONNECT;
+	}
+	// Don't use "default" case. This should generate a warning if we ever add any verbs. But
+	// still assert here for safety.
+	assert(false);
+	return GET;
+}
+
+HttpSession::HttpSession(const HttpClient &client, events::EventLoop &event_loop) :
+	logger_("http"),
+	resolver_(GetAsioIoContext(event_loop)),
+	stream_(GetAsioIoContext(event_loop)),
+	body_buffer_(16384) {
+	response_buffer_.reserve(body_buffer_.size());
+}
+
+HttpSession::~HttpSession() {
+	Cancel();
+}
+
+error::Error HttpSession::AsyncCall(
+	HttpRequestPtr req, HttpResponseHandler header_handler, HttpResponseHandler body_handler) {
+	Cancel();
+
+	if (!req->ready_) {
+		return error::MakeError(error::ProgrammingError, "HttpRequest is not ready");
+	}
+
+	if (!header_handler || !body_handler) {
+		return error::MakeError(
+			error::ProgrammingError, "header_handler and body_handler can not be nullptr");
+	}
+
+	if (req->protocol_ != "http") {
+		return error::Error(make_error_condition(errc::protocol_not_supported), req->protocol_);
+	}
+
+	// Use url as context for logging.
+	logger_ = log::Logger("http").WithFields(log::LogField("url", req->orig_address_));
+
+	request_ = req;
+	header_handler_ = header_handler;
+	body_handler_ = body_handler;
+
+	resolver_.async_resolve(
+		request_->host_,
+		to_string(request_->port_),
+		[this](error_code err, const asio::ip::tcp::resolver::results_type &results) {
```

Comment:
Q: does this capture `this` by copy? :thinking:

Reply:
Yes, but `this` is a pointer to the object, not the object itself.